src/janus.c

```diff
@@ -899,20 +899,32 @@ static void janus_request_free(const janus_refcount *request_ref) {
 		janus_refcount_decrease(&request->instance->ref);
 	request->instance = NULL;
 	request->request_id = NULL;
-	if(request->message)
+	if(request->message) {
 		json_decref(request->message);
-	request->message = NULL;
+		request->message = NULL;
+	}
+	if(request->error) {
+		g_free(request->error);
+		request->error = NULL;
+	}
 	g_free(request);
 }
 
-janus_request *janus_request_new(janus_transport *transport, janus_transport_session *instance, void *request_id, gboolean admin, json_t *message) {
+janus_request *janus_request_new(janus_transport *transport, janus_transport_session *instance, void *request_id, gboolean admin, json_t *message, json_error_t *error) {
 	janus_request *request = g_malloc(sizeof(janus_request));
 	request->transport = transport;
 	request->instance = instance;
 	janus_refcount_increase(&instance->ref);
 	request->request_id = request_id;
 	request->admin = admin;
 	request->message = message;
+	if(error) {
+		request->error = (json_error_t*)g_malloc(sizeof(json_error_t));
+		*request->error = *error;
```

Comment:
I don't understand this. You're first allocating memory for `request->error`, but then just have the pointer point to an existing memory area?

Reply:
I think the intent is to initialize the struct pointed to by `request->error` using the contents of `error` (e.g., *dst = *src)...I have some vague memory that this isn't safe on all platforms (maybe depending on padding/alignment?) but I can't find a definitive reference to support that:

Maybe not the most rigorous test but in this example `memcpy` and `dereferencing` lead to the same asm: 

https://godbolt.org/z/5x69qsqaG

Reply:
> I think the intent is to initialize the struct pointed to by `request->error` using the contents of `error` (e.g., *dst = *src).

Exactly, I'm copying the struct from the incoming error struct (on the stack in this case) to the instance store here (on the heap as it may or may not be set). 

I'd be really surprised if there's a platform problem with assigning one struct to another. I verified that the error struct uses fixed sized arrays so the copy should be, as you point out, equivalent to a memcpy. =)

Open to any changes desired here of course.

Reply:
@lminiero - maybe you missed the deference of `request->error` on that line?

Reply:
I had seen it, but not knowing what `json_error_t` is I just assumed it was a string, rather than a struct. Maybe an explicit memcpy is a safer approach, even though I think we use the struct copy assignment ourselves around Janus (e.g., for RTP extensions) so if that's broken somewhere, we're in trouble anyway :hand_over_mouth:

Reply:
Yeah I think you can disregard what I said, structure copy by assignment is fine (if not better since the compiler can make smarter optimizations re: padding) in C90 (which Janus already requires IIRC), I think my memory of having to use `memcpy` was for some niche case.